node/src/isolate/external.cpp

```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
     std::atomic<id_t> cur_id;
     std::shared_ptr<dispatch<io::context_tag>> signal_dispatch;
     bool prepared;
-    bool connecting;
 
     std::chrono::time_point<std::chrono::system_clock> last_failed_connect_time;
     std::chrono::milliseconds seal_time;
+
     static constexpr std::chrono::milliseconds min_seal_time {1000ul};
     static constexpr std::chrono::milliseconds max_seal_time {1000000ul};
 
     std::vector<std::shared_ptr<spool_load_t>> spool_queue;
     std::vector<std::shared_ptr<spawn_load_t>> spawn_queue;
 
+    std::atomic_flag connecting_ = ATOMIC_FLAG_INIT;
+
+    // All async methods should capture shared_from_this()
+    struct connection_t : std::enable_shared_from_this<connection_t> {
+
+        using async_completion = std::function<void(const std::error_code&, std::unique_ptr<tcp::socket>)>;
+
+        const std::shared_ptr<inner_t> parent_;
+        logging::logger_t& log_;
+
+        std::unique_ptr<tcp::socket> socket_;
+        asio::deadline_timer connect_timer_;
+
+        connection_t(const std::shared_ptr<inner_t>& parent, logging::logger_t &log, asio::io_service& io):
```

Comment:
Let's follow the style for constructors:
```cpp
class_t(arg1, arg2, arg3)
: member1(arg1)
, member2(arg2)
, member3(arg3) {
}
```

Also: `logging::logger_t &log` -> `logging::logger_t& log`